index.d.ts

```diff
@@ -0,0 +1,373 @@
+// TypeScript Version: 4.2
+
+export type ImporterReturnType =
+  | {file: string}
+  | {contents: string}
+  | Error
+  | null;
+
+export type Importer = (
+  url: string,
+  prev: string,
+  done: (data: ImporterReturnType) => void
+) => ImporterReturnType | void;
+
+export interface LogData {
+  /**
+   * The message as-is in the file/data
+   */
+  message: string;
+
+  /**
+   * A set of information related to the location of the error
+   */
+  source: SourceSpan;
+
+  /**
+   * Is it a deprecation message
+   */
+  deprecation: boolean;
+}
+
+export interface Logger {
+  /**
+   * A function that is called each time `@warn` is encountered
+   * @param data All data relative to the `@warn` line
+   */
+  warn?(data: LogData): void;
+
+  /**
+   * A function that is called each time `@debug` is encountered
+   * @param data All data relative to the `@debug` line
+   */
+  debug?(data: LogData): void;
+}
+
+export interface Options {
+  /**
+   * Path to a file to compile.
+   *
+   * @default null
+   */
+  file?: string;
+
+  /**
+   * A string to pass to compile.
+   *
+   * It is recommended that you use `includePaths` in conjunction with this so that sass can find files when using the @import directive.
+   *
+   * @default null
+   */
+  data?: string;
+
+  /**
+   * Handles when the @import directive is encountered.
+   *
+   * A custom importer allows extension of the sass engine in both a synchronous and asynchronous manner.
+   *
+   * @default undefined
+   */
+  importer?: Importer | Importer[];
+
+  /**
+   * Holds a collection of custom functions that may be invoked by the sass files being compiled.
+   *
+   * @default undefined
+   */
+  functions?: {
+    [key: string]: (...args: types.SassType[]) => types.SassType | void;
+  };
```

Comment:
Like importers, functions can be either synchronous or asynchronous. If they're asynchronous, the last argument is a `done()` callback.